### ISSUE TITLE
tensorflow version of this code

### ISSUE THREAD

**Issue by whuhxb** [lnsmith54/CFL#2]
Hi,
   Is there a tensorflow version of this code?
Thanks.

**Comment by lnsmith54** [comment]
There isn't a tensorflow version of this code but there are tensorflow versions of the focal loss.  It should be fairly easy to modify those codes to perform CFL.  Not so easy will be to thoroughly test it.

**Comment by whuhxb** [comment]
OK Thanks

**Comment by whuhxb** [comment]
@lnsmith54 
Hi, I have revised it into tensorflow version. Could you please help double check?
Thanks.

**Comment by whuhxb** [comment]
@lnsmith54 
The revised .py file is as below.

#  https://github.com/Alibaba-MIIL/ASL
import tensorflow as tf


class AsymmetricLoss:
    def __init__(self, gamma_neg=4, gamma_pos=1, clip=0.05, eps=1e-8, disable_torch_grad_focal_loss=True):
        #super(AsymmetricLoss, self).__init__()

        self.gamma_neg = gamma_neg
        self.gamma_pos = gamma_pos
        self.clip = clip
        self.disable_torch_grad_focal_loss = disable_torch_grad_focal_loss
        self.eps = eps

    def forward(self, x, y):
        """"
        Parameters
        ----------
        x: input logits
        y: targets (multi-label binarized vector)
        """

        # Calculating Probabilities
        x_sigmoid = tf.sigmoid(x)
        xs_pos = x_sigmoid
        xs_neg = 1 - x_sigmoid

        # Asymmetric Clipping
        if self.clip is not None and self.clip > 0:
            xs_neg = (xs_neg + self.clip).clip_by_value(max=1)

        # Basic CE calculation
        los_pos = y * tf.log(xs_pos.clip_by_value(min=self.eps))
        los_neg = (1 - y) * tf.log(xs_neg.clip_by_value(min=self.eps))
        loss = los_pos + los_neg

        # Asymmetric Focusing
        if self.gamma_neg > 0 or self.gamma_pos > 0:
            #if self.disable_torch_grad_focal_loss:
            #    torch.set_grad_enabled(False) # how?
            pt0 = xs_pos * y
            pt1 = xs_neg * (1 - y)  # pt = p if t > 0 else 1-p
            pt = pt0 + pt1
            one_sided_gamma = self.gamma_pos * y + self.gamma_neg * (1 - y)
            one_sided_w = tf.pow(1 - pt, one_sided_gamma)
            #if self.disable_torch_grad_focal_loss:
            #    torch.set_grad_enabled(True) # how?
            loss *= one_sided_w

        return -tf.reduce_sum(loss)


class AsymmetricLossOptimized:
    ''' Notice - optimized version, minimizes memory allocation and gpu uploading,
    favors inplace operations'''

    def __init__(self, gamma_neg=4, gamma_pos=1, clip=0.05, eps=1e-8, disable_torch_grad_focal_loss=False):
        #super(AsymmetricLossOptimized, self).__init__()

        self.gamma_neg = gamma_neg
        self.gamma_pos = gamma_pos
        self.clip = clip
        self.disable_torch_grad_focal_loss = disable_torch_grad_focal_loss
        self.eps = eps

        # prevent memory allocation and gpu uploading every iteration, and encourages inplace operations
        self.targets = self.anti_targets = self.xs_pos = self.xs_neg = self.asymmetric_w = self.loss = None

    def forward(self, x, y):
        """"
        Parameters
        ----------
        x: input logits
        y: targets (multi-label binarized vector)
        """

        self.targets = y
        self.anti_targets = 1 - y

        # Calculating Probabilities
        self.xs_pos = tf.sigmoid(x)
        self.xs_neg = 1.0 - self.xs_pos

        # Asymmetric Clipping
        if self.clip is not None and self.clip > 0:
            self.xs_neg.add_(self.clip).clip_by_value(max=1)

        # Basic CE calculation
        self.loss = self.targets * tf.log(self.xs_pos.clip_by_value(min=self.eps))
        #self.loss.add_(self.anti_targets * tf.log(self.xs_neg.clip_by_value(min=self.eps)))
        self.loss += self.anti_targets * tf.log(self.xs_neg.clip_by_value(min=self.eps))

        # Asymmetric Focusing
        if self.gamma_neg > 0 or self.gamma_pos > 0:
            #if self.disable_torch_grad_focal_loss:
            #    torch.set_grad_enabled(False)
            self.xs_pos = self.xs_pos * self.targets
            self.xs_neg = self.xs_neg * self.anti_targets
            self.asymmetric_w = tf.pow(1 - self.xs_pos - self.xs_neg,
                                          self.gamma_pos * self.targets + self.gamma_neg * self.anti_targets)
            #if self.disable_torch_grad_focal_loss:
            #    torch.set_grad_enabled(True)
            self.loss *= self.asymmetric_w

        return -tf.reduce_sum(self.loss)


class ASLSingleLabel:
    '''
    This loss is intended for single-label classification problems
    '''
    def __init__(self, gamma_pos=0, gamma_neg=4, eps: float = 0.1, reduction='mean'):
        #super(ASLSingleLabel, self).__init__()

        self.eps = eps
        self.logsoftmax = tf.nn.LogSoftmax(dim=-1)
        self.targets_classes = []
        self.gamma_pos = gamma_pos
        self.gamma_neg = gamma_neg
        self.reduction = reduction
#        print("ASLSingleLabel: gamma_pos=", gamma_pos," gamma_neg=",gamma_neg, " eps=",eps)

    def forward(self, inputs, target):
        '''
        "input" dimensions: - (batch_size,number_classes)
        "target" dimensions: - (batch_size)
        '''
        num_classes = inputs.size()[-1]
        log_preds = self.logsoftmax(inputs)
        if len(list(target.size()))>1:
            target = tf.argmax(target, 1)
        #self.targets_classes = tf.zeros_like(inputs).scatter_(1, target.long().unsqueeze(1), 1)
        self.targets_classes = tf.one_hot(tf.expand_dims(target,1), num_classes)
        print ('self.targets_classes', self.targets_classes) # added

        # ASL weights
        targets = self.targets_classes
        anti_targets = 1 - targets
        xs_pos = tf.exp(log_preds)
        xs_neg = 1 - xs_pos
        xs_pos = xs_pos * targets
        xs_neg = xs_neg * anti_targets
        asymmetric_w = tf.pow(1 - xs_pos - xs_neg,
                                 self.gamma_pos * targets + self.gamma_neg * anti_targets)
        log_preds = log_preds * asymmetric_w

        if self.eps > 0:  # label smoothing
            #self.targets_classes = self.targets_classes.mul(1 - self.eps).add(self.eps / num_classes)
            self.targets_classes = tf.add(tf.multiply(self.targets_classes, 1 - self.eps), self.eps / num_classes)

        # loss calculation
        #loss = - self.targets_classes.mul(log_preds)
        loss = - tf.multiply(self.targets_classes, log_preds)

        #loss = loss.sum(dim=-1)
        loss = tf.reduce_sum(loss) # loss = tf.reduce_sum(loss, axis=1) 
        if self.reduction == 'mean':
            loss = tf.reduce_mean(loss)

        return loss

class Cyclical_FocalLoss:
    '''
    This loss is intended for single-label classification problems
    '''
    def __init__(self, gamma_pos=0, gamma_neg=4, gamma_hc=0, eps: float = 0.1, reduction='mean', epochs=200, 
                 factor=2):
        #super(Cyclical_FocalLoss, self).__init__()

        self.eps = eps
        self.logsoftmax = tf.nn.LogSoftmax(dim=-1)
        self.targets_classes = []
        self.gamma_hc = gamma_hc
        self.gamma_pos = gamma_pos
        self.gamma_neg = gamma_neg
        self.reduction = reduction
        self.epochs = epochs
        self.factor = factor # factor=2 for cyclical, 1 for modified
#        self.ceps = ceps
#        print("Asymetric_Cyclical_FocalLoss: gamma_pos=", gamma_pos," gamma_neg=",gamma_neg, 
#              " eps=",eps, " epochs=", epochs, " factor=",factor)

    def forward(self, inputs, target, epoch):
        '''
        "input" dimensions: - (batch_size,number_classes)
        "target" dimensions: - (batch_size)
        '''
#        print("input.size(),target.size()) ",inputs.size(),target.size())
        num_classes = inputs.size()[-1]
        log_preds = self.logsoftmax(inputs)
        if len(list(target.size()))>1:
            target = tf.argmax(target, 1)
        self.targets_classes = tf.one_hot(tf.expand_dims(target,1), num_classes)
        print ('self.targets_classes', self.targets_classes) # added

        # Cyclical
#        eta = abs(1 - self.factor*epoch/(self.epochs-1))
        if self.factor*epoch < self.epochs:
            eta = 1 - self.factor *epoch/(self.epochs-1)
        elif self.factor == 1.0:
            eta = 0
        else:
            eta = (self.factor*epoch/(self.epochs-1) - 1.0)/(self.factor - 1.0) 

        # ASL weights
        targets = self.targets_classes
        anti_targets = 1 - targets
        xs_pos = tf.exp(log_preds)
        xs_neg = 1 - xs_pos
        xs_pos = xs_pos * targets
        xs_neg = xs_neg * anti_targets
        asymmetric_w = tf.pow(1 - xs_pos - xs_neg,
                                 self.gamma_pos * targets + self.gamma_neg * anti_targets)
        positive_w = tf.pow(1 + xs_pos,self.gamma_hc * targets)
        log_preds = log_preds * ((1 - eta)* asymmetric_w + eta * positive_w)

        if self.eps > 0:  # label smoothing
            self.targets_classes = tf.add(tf.multiply(self.targets_classes, 1 - self.eps), self.eps / num_classes)


        # loss calculation
        loss = - tf.multiply(self.targets_classes, log_preds)

        loss = tf.reduce_sum(loss) # loss = tf.reduce_sum(loss, axis=1) 
        if self.reduction == 'mean':
            loss = tf.reduce_mean(loss)

        return loss

class ASL_FocalLoss:
    '''
    This loss is intended for single-label classification problems
    '''
    def __init__(self, gamma=2, eps: float = 0.1, reduction='mean'):
        #super(ASL_FocalLoss, self).__init__()

        self.eps = eps
        self.logsoftmax = tf.nn.LogSoftmax(dim=-1)
        self.targets_classes = []
        self.gamma = gamma
        self.reduction = reduction
#        print("ASL_FocalLoss: gamma=", gamma, " eps=",eps)

    def forward(self, inputs, target):
        '''
        "input" dimensions: - (batch_size,number_classes)
        "target" dimensions: - (batch_size)
        '''
        num_classes = inputs.size()[-1]
        log_preds = self.logsoftmax(inputs)
        self.targets_classes = tf.one_hot(tf.expand_dims(target,1), num_classes)
        print ('self.targets_classes', self.targets_classes) # added

        # ASL weights
        targets = self.targets_classes
        anti_targets = 1 - targets
        xs_pos = tf.exp(log_preds)
        xs_neg = 1 - xs_pos
        xs_pos = xs_pos * targets
        xs_neg = xs_neg * anti_targets
        asymmetric_w = tf.pow(1 - xs_pos - xs_neg,
                                 self.gamma * targets + self.gamma * anti_targets)
        log_preds = log_preds * asymmetric_w

        if self.eps > 0:  # label smoothing
            self.targets_classes = tf.add(tf.multiply(self.targets_classes, 1 - self.eps), self.eps / num_classes)


        # loss calculation
        loss = - tf.multiply(self.targets_classes, log_preds)

        loss = tf.reduce_sum(loss) # loss = tf.reduce_sum(loss, axis=1) 
        if self.reduction == 'mean':
            loss = tf.reduce_mean(loss)

        return loss
